### PR TITLE
Simulated Scaling Documentation

### PR DESCRIPTION
**Pull request by TylerJewell**
Adds a section to the managing doc of the admin-guide to discuss how to use Docker Machine to simulate running a 3 node Codenvy cluster where 2 of the nodes are workspace node and the third is a Codenvy node.

Need to get more feedback from @garagatyi on my last open comments, which is why we are having admins modify codenvy.env to setup the cluster instead of using the codenvy add-node command which is the prescribed way.

### CODE REVIEW

File: docs/_docs/admin-guide/admin-managing.md [hunk 1]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+Codenvy requires a Docker overlay network to exist for our workspace nodes. An overlay network is a network that spans across the various nodes that allows Docker containers to simplify how they communicate with one another. This is mandatory for Codenvy since your workspaces can themselves be composed of multiple containers (such as defined by Docker Compose). If a single workspace has multiple runtimes, we can deploy those runtimes on different physical nodes. An overlay network allows those containers to have a common nework so that they can communicate with each other using container names, without each container having to understand the location of the other.
 
-You can remove nodes with `codenvy remove-node <ip>`.
+Overlay networks require a distributed key-value store to be running on a node. We embed Consul, a key-value storage implementation as part of the Codenvy master node. We currently only support adding Linux nodes into an overlay network.
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: I think that key-value storage can not be on a docker node for workspaces. But I read these docs as it should be there. Can you rephrase that sentence.

File: docs/_docs/admin-guide/admin-managing.md [hunk 2]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+
+The default network in Docker is a "bridge" network. If you know that your users will only ever have single container workspaces (this would be unusual and rare), then you can continue using the bridge network for scaling. Bridge networks can be scaled with Linux, Windows, or Mac.
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: We don't support more than 1 node with a bridge networking. So it is not scaling if user can't start new node on that network.

File: docs/_docs/admin-guide/admin-managing.md [hunk 3]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+# Codenvy IP is either set by you to CODENVY_HOST or auto-discovered with:
+docker run --net=host eclipse/che-ip:nightly
+
+# Get the network interface for your ws node, typically 'eth1' or 'eth0':
+ifconfig
+```
+
+2: On the Codenvy master node, start Consul: `docker -H <CODENVY-IP>:2376 run -d -p 8500:8500 -h consul progrium/consul -server -bootstrap`
+
+3: On each workspace node, [configure and restart Docker](https://docs.docker.com/engine/admin/) with new options: `--cluster-store=consul://<CODENVY-IP>:8500`, `--cluster-advertise=<WS-IF>:2376`, and `--engine-insecure-registry=<CODENVY-IP>:5000`.
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: They have to add listening on TCP port into docker daemon conf also.

File: docs/_docs/admin-guide/admin-managing.md [hunk 4]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+
+4: Verify that Docker is running properly. Docker will not start if it is not able to connect to the key-value storage. TODO: ADD TEST INSTRUCTION. Each workspace node that successfully runs this command is part of the overlay network.
+
+5: On the Codenvy master node, modify `codenvy.env` to uncomment or add:
+```json
+# Comma-separated list of IP addresses for each workspace node
+CODENVY_SWARM_NODES=<WS-IP>:2376,<WS2-IP>:2376,<WSn-IP>:2376
+```
+
+6: Restart Codenvy with `codenvy/cli restart`.
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: Roman said that we can reload just swarm container. So you can test it with him. It is better way if it works.

File: docs/_docs/admin-guide/admin-managing.md [hunk 5]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+
+This simulated scaling can be used for production, but it is generally discouraged because you would be running Docker in VMs that are on a host, and you are just taking on some extra I/O overhead that may not generally be necessary.  However, this simulated-based approach gives good pointers on configuration of a distributed, cluster-based system if you were to use VMs-only.
+
+As an example, the following sequence launches a 3-node cluster of Codenvy using Docker machine with a VirtualBox hypervisor. In this example, we launch 4 VMs: a Codenvy node, 2 additional workspace nodes, and a node to handle key-value storage. The key-value storage node is typically not part of the scaling configuration. However, Codenvy requires an "overlay" network, which is powered by a key-value storage provider such as Consule, etcd, or zookeeper. When running Codenvy on the host, we are able to setup an etcd key-value storage system automatically and associate the nodes with it. However, in a VM scale-out scenario, a dedicated key-value storage provider is needed. This particular example uses Consul key-value storage to setup the overlay network.
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: > When running Codenvy on the host, we are able to setup an etcd key-value storage system automatically and associate the nodes with it

Is it about dockerized Codenvy or not? In dockerized Codenvy we don't install KV storage yet. Maybe we should not put  that in docs for now?
If it is about non dockerized Codenvy I would recommend to specify that.

File: docs/_docs/admin-guide/admin-managing.md [hunk 6]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+docker -H <KV-IP:2376> run -d -p 8500:8500 -h consul progrium/consul -server -bootstrap
+```
+
+Start 3 VMs named 'codenvy', 'ws1', 'ws2'):
+```shell
+# Codenvy 
+# Grab the IP address of this VM and use it in other commands where we have <CODENVY-IP>
+docker-machine create -d virtualbox --engine-env DOCKER_TLS=no --virtualbox-memory "2048" \
+                      --engine-opt host=tcp://0.0.0.0:2375 \
+                      --engine-opt="cluster-store=consul://<KV-IP>:8500" \
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: I think that for codenvy master node which doesn't take part in workspace infrastructure we don't need cluster config

File: docs/_docs/admin-guide/admin-managing.md [hunk 7]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+                      --engine-opt="cluster-advertise=eth1:2376" codenvy
+
+# Workspace Node 1
+docker-machine create -d virtualbox --engine-env DOCKER_TLS=no --virtualbox-memory "3000" \
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: Maybe we should mention that 2048 is default WS size, so this node will be able to run single WS only. And we can use <3000 MB in that case.

File: docs/_docs/admin-guide/admin-managing.md [hunk 8]
```diff
@@ -6,17 +6,104 @@ layout: docs
...
+```shell
+# SSH into the VM
+docker-machine ssh codenvy
+
+# Initialize a Codenvy installation
+docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock \
+           -v /home/docker/.codenvy:/data codenvy/cli:nightly init
+
+# Setup Codenvy's configuration file to have the IP addresses of each workspace node
+sed -i "s/^#CODENVY_WORKSPACE_AUTO_SNAPSHOT=true.*/CODENVY_WORKSPACE_AUTO_SNAPSHOT=true/g" ~/.codenvy/codenvy.env
```
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: Add sudo before seds.

File: docs/_docs/admin-guide/admin-managing.md (hunk 4 above)
Comment (TylerJewell) [inline docs/_docs/admin-guide/admin-managing.md]: We do not have that syntax in the CLI.  Until the CLI supports such a command, we only have restart.

File: docs/_docs/admin-guide/admin-managing.md (hunk 5 above)
Comment (TylerJewell) [inline docs/_docs/admin-guide/admin-managing.md]: It is only for dockerized codenvy.  The simulated scaling example runs consul kv storage, so we are covered.

File: docs/_docs/admin-guide/admin-managing.md (hunk 6 above)
Comment (TylerJewell) [inline docs/_docs/admin-guide/admin-managing.md]: You added the cluster configuration to your example.  Can you update your example to indicate which of the 4 options are needed in the Docker daemon (on the Codenvy host) if the Codenvy master **does not** need to have workspaces?

There are four - which ones do we drop?

`--cluster-store=consul://<CODENVY-IP>:8500`, `--cluster-advertise=<WS-IF>:2376`, '--host=tcp://0.0.0.0:2375`, and `--engine-insecure-registry=<CODENVY-IP>:5000`

File: docs/_docs/admin-guide/admin-managing.md (hunk 8 above)
Comment (TylerJewell) [inline docs/_docs/admin-guide/admin-managing.md]: Your example did not have sudo in the other docs - why is it needed now.

**Comment by TylerJewell** [comment]
I've incorporated most comments - just need instructions about which parameters are **not** required on the master Daemon configuration if the master is not going to host workspaces.

File: docs/_docs/admin-guide/admin-managing.md (hunk 8 above)
Comment (garagatyi) [inline docs/_docs/admin-guide/admin-managing.md]: I updated example yesterday with sudo

**Comment by TylerJewell** [comment]
This issue is only about documenting how the systm works today. Other
issues are related to improving how the system operates, and this sort of
recycling is part of that.

Tyler Jewell | CEO | tyler@​codenvy.​com | 9​78​.8​84​.53​55
